xxxx-ledger-plugin-interface-2/xxxx-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,502 @@
+
+Rejects with `InvalidFieldsError` if required fields are missing from the transfer or malformed. Rejects with `NotAcceptedError` if the transfer is rejected by the ledger due to insufficient balance or any other reason. Rejects with `NotConnectedError` if the plugin is not connected to the ledger.
+
+Rejects with [`InterledgerRejectionError`](#class-interledgerrejectionerror) if the other side rejects the transfer and attaches ILP rejection data. Rejects with `GenericRejectionError` if the other side rejects the transfer, but does not attach valid ILP rejection data.
+
+This method MAY reject with any arbitrary JavaScript error.
+
+###### Example
+```js
+p.sendTransfer({
+  amount: '10',
+  destination: 'example.us.acmebank.bob',
+  executionCondition: '47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU',
+  expiresAt: '2016-05-18T12:00:00.000Z',
+  data: Buffer.alloc(0),
+  custom: {}
+})
+```
+
+For a detailed description of these properties, please see [`Class: Transfer`](#class-transfer).
+
+#### LedgerPlugin#registerTransferHandler
+<code>ledgerPlugin.registerTransferHandler( **transferHandler**: ( transfer: [Transfer](#class-transfer) ) ⇒ Promise&lt;[Message](#class-message)> ) ⇒ null</code>
+
+Set the callback which is used to handle incoming prepared transfers. The callback should expect one parameter (the [Transfer](#class-transfer)) and return a promise for the resulting [FulfillmentInfo](#class-fulfillmentinfo). If the transfer is rejected or an error occurs, the callback should reject the transfer. In general, the callback should behave as [`sendTransfer`](#ledgerpluginsendtransfer) does.
+
+If a transfer handler is already set, this method throws a `TransferHandlerAlreadyRegisteredError`. In order to change the transfer handler, the old handler must first be removed via [`deregisterTransferHandler`](#ledgerpluginderegistertransferhandler). This is to ensure that handlers are not overwritten by accident.
+
+If an incoming transfer is received by the plugin, but no handler is registered, the plugin should reject the transfer.
+
+#### LedgerPlugin#deregisterTransferHandler
+<code>ledgerPlugin.deregisterTransferHandler( ) ⇒ null</code>
+
+Removes the currently used transfer handler. This has the same effect as if [`registerTransferHandler`](#ledgerpluginregistertransferhandler) had never been called.
+
+If no transfer handler is currently set, this method does nothing.
+
+### Event: `info_change`
+<code style="">ledgerPlugin.on('info_change',
+  (
+    **info**:[LedgerInfo](#class-ledgerinfo)
+  ) ⇒
+)</code>
+
+Emitted any time the plugin's `LedgerInfo` cache changes.
+
+## Class: Transfer
+<code>class Transfer</code>
+
+The `Transfer` class is used to describe transfers from the originator of the sendTransfer call towards some `destination`. All fields are required and MUST NOT be undefined. However, `amount` MAY be the value `'0'`, `data` MAY be an empty buffer and `custom` MAY be an empty object.
+
+###### Fields
```

Comment:
In the light of https://github.com/interledger/rfcs/pull/349 I would change this to:
```js
{
  headers: {
    'ILP-Destination': 'g.crypto.bitcoin.1XPTgDRhN8RFnzniWCddobD9iKZatrvH4.~asdf1234',
    'ILP-Condition': 'x73kz0AGyqYqhw_c5LqMhSgpcOLF3rBS8GdR52hLpB8',
    'ILP-Expiry': '2017-12-07T18:47:59.015Z',
    'ILP-Amount': '1000'
 },
 body: <body>
}
```

Reply:
Or maybe
```js
{
  destination: 'g.crypto.bitcoin.1XPTgDRhN8RFnzniWCddobD9iKZatrvH4.~asdf1234',
  condition: 'x73kz0AGyqYqhw_c5LqMhSgpcOLF3rBS8GdR52hLpB8',
  expiry: '2017-12-07T18:47:59.015Z',
  amount: '1000'
  body: <body>
}
```

Reply:
-1 for the first `transfer` proposal, but +1 for something like the 2nd `transfer` proposal.

* The prefixes (i.e., `ILP-...` in #349 are there as an artifact of using HTTP headers to carry the data (to distinguish between other headers that might be arbitrarily added to the request as part of a normal and valid HTTP request). Conversely, In this interface, the prefixes are not needed because there's less of a risk of "other data fields" colliding with these fields names -- mainly because we're defining the field names here.
* `body` is an artifact of HTTP. We should call it what it really is here (I like Adrian's suggestion above of `peerData` and `pluginData`), and just make new attributes for the data we need.